src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,460 @@
+        }
+
+        // Not all gpus provide an 'efficient' frequency signal, and the
+        // value provided by the policy may not be valid.  In this case approximating
+        // f_efficient as midway between F_min and F_max is reasonable.
+        if (std::isnan(in_policy[M_POLICY_GPU_FREQ_EFFICIENT])) {
+            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = (in_policy[M_POLICY_GPU_FREQ_MAX]
+                                                             + accel_min_freq) / 2;
+        }
+
+        if (in_policy[M_POLICY_GPU_FREQ_EFFICIENT] > accel_max_freq ||
+            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] < accel_min_freq ) {
+            throw Exception("GPUActivityAgent::" + std::string(__func__) +
+                            "(): GPU_FREQ_EFFICIENT out of range: " +
+                            std::to_string(in_policy[M_POLICY_GPU_FREQ_EFFICIENT]) +
+                            ".", GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if (in_policy[M_POLICY_GPU_FREQ_EFFICIENT] > in_policy[M_POLICY_GPU_FREQ_MAX]) {
+            throw Exception("GPUActivityAgent::" + std::string(__func__) +
+                            "(): GPU_FREQ_EFFICIENT (" +
+                            std::to_string(in_policy[M_POLICY_GPU_FREQ_EFFICIENT]) +
+                            ") value exceeds GPU_FREQ_MAX (" +
+                            std::to_string(in_policy[M_POLICY_GPU_FREQ_MAX]) +
+                            ").", GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        // If no phi value is provided assume the default behavior.
+        if (std::isnan(in_policy[M_POLICY_GPU_PHI])) {
+            in_policy[M_POLICY_GPU_PHI] = M_POLICY_PHI_DEFAULT;
+        }
+
+        if (in_policy[M_POLICY_GPU_PHI] < 0.0 ||
+            in_policy[M_POLICY_GPU_PHI] > 1.0) {
+            throw Exception("GPUActivityAgent::" + std::string(__func__) +
+                            "(): POLICY_GPU_PHI value out of range: " +
+                            std::to_string(in_policy[M_POLICY_GPU_PHI]) + ".",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        // Policy provided initial values
+        double f_max = in_policy[M_POLICY_GPU_FREQ_MAX];
+        double f_efficient = in_policy[M_POLICY_GPU_FREQ_EFFICIENT];
+        double phi = in_policy[M_POLICY_GPU_PHI];
+
+        // initial range is needed to apply phi
+        double f_range = f_max - f_efficient;
+
+        if (phi > 0.5) {
+            //Energy Biased.  Scale F_max down to F_efficient based upon phi value
+            //Active region phi usage
+            f_max = std::max(f_efficient, f_max - f_range * (phi-0.5) / 0.5);
+        }
+        else if (phi < 0.5) {
+            //Perf Biased.  Scale F_efficient up to F_max based upon phi value
+            //Active region phi usage
+            f_efficient = std::min(f_max, f_efficient + f_range * (0.5-phi) / 0.5);
+        }
```

Comment:
What about when `phi = 0.5`?

Reply:
When phi=0.5 we allow the agent to steer frequency in the range of Fe to Fmax without modifying either Fe or Fmax.